spring-boot/admin/src/main/java/eu/coatrack/admin/service/GatewayHealthMonitorService.java

```diff
@@ -0,0 +1,132 @@
+                .stream()
+                .anyMatch(gateway -> gateway.status == ProxyStates.CRITICAL)) {
+            return ProxyStates.CRITICAL;
+        } else if (gatewaysWithMonitoringActivatedList
+                .stream()
+                .anyMatch(gateway -> gateway.status == ProxyStates.WARNING)) {
+            return ProxyStates.WARNING;
+        } else if (gatewaysWithMonitoringActivatedList.stream().allMatch(gateway -> gateway.status == ProxyStates.OK)) {
+            return ProxyStates.OK;
+        }
+        return ProxyStates.NEVER_CONNECTED;
+    }
+
+    public List<GatewayDataForTheGatewayHealthMonitor> updateProxyInfoForGatewayHealthMonitor() {
+        List<GatewayDataForTheGatewayHealthMonitor> gatewayDataForGatewayHealthMonitorList = new ArrayList<>();
+        List<Proxy> allProxiesOwnedByTheLoggedInUser = proxyRepository.findAvailable();
+        allProxiesOwnedByTheLoggedInUser.forEach((proxy) -> {
+            GatewayDataForTheGatewayHealthMonitor gatewayDataForGatewayHealthMonitor = new GatewayDataForTheGatewayHealthMonitor();
+            gatewayDataForGatewayHealthMonitor.setDataNeededFromProxy(proxy);
+            if (proxy.isMonitoringEnabled()) {
+                if (proxy.getTimeOfLastSuccessfulCallToAdmin() != null) {
+                    gatewayDataForGatewayHealthMonitor.minutesPastSinceLastContact = Duration.between(proxy.getTimeOfLastSuccessfulCallToAdmin(), LocalDateTime.now()).toMinutes();
+                    if (proxy.getTimeOfLastSuccessfulCallToAdmin()
+                            .plusMinutes(gatewayHealthCriticalThresholdInMinutes)
+                            .isBefore(LocalDateTime.now())) {
+                        gatewayDataForGatewayHealthMonitor.status = ProxyStates.CRITICAL;
+                    } else if (proxy.getTimeOfLastSuccessfulCallToAdmin()
+                            .plusMinutes(gatewayHealthWarningThresholdInMinutes)
+                            .isBefore(LocalDateTime.now())) {
+                        gatewayDataForGatewayHealthMonitor.status = ProxyStates.WARNING;
+                    } else {
+                        gatewayDataForGatewayHealthMonitor.status = ProxyStates.OK;
+                    }
+                } else {
+                    gatewayDataForGatewayHealthMonitor.status = ProxyStates.NEVER_CONNECTED;
+                }
+            } else {
+                gatewayDataForGatewayHealthMonitor.status = ProxyStates.IGNORE;
+            }
+            gatewayDataForGatewayHealthMonitorList.add(gatewayDataForGatewayHealthMonitor);
+        });
+        Collections.sort(gatewayDataForGatewayHealthMonitorList, new GatewayHealthDataComparator());
+        return gatewayDataForGatewayHealthMonitorList;
+    }
+
+    class GatewayHealthDataComparator implements Comparator<GatewayDataForTheGatewayHealthMonitor> {
+        @Override
+        public int compare(GatewayDataForTheGatewayHealthMonitor a, GatewayDataForTheGatewayHealthMonitor b) {
```

Comment:
Maybe add a comment to explain the "ordering rules".

Reply:
Done